climada/util/lines_polys_handler.py

```diff
@@ -0,0 +1,886 @@
+terms of the GNU Lesser General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+"""
+import logging
+import copy
+
+import geopandas as gpd
+import pandas as pd
+import numpy as np
+import shapely as sh
+import scipy as sp
+import shapely.geometry as shgeom
+import cartopy.crs as ccrs
+
+from shapely.ops import unary_union
+
+from climada.engine import Impact
+from climada.util import coordinates as u_coord
+
+import pyproj
+
+LOGGER = logging.getLogger(__name__)
+
+
+def calc_geom_impact(
+        exp, impf_set, haz, res,
+        to_meters=False, disagg=None, agg='sum'
+        ):
+    """
+    Compute impact for exposure with (multi-)polygons and/or (multi-)lines.
+    Lat/Lon values in exp.gdf are ignored, only exp.gdf.geometry is considered.
+
+    The geometries are first disaggrated to a grid with resolution res*res.
+    The impact per point is then aggregated for each geometry.
+
+    Parameters
+    ----------
+    exp : Exposures
+        The exposure instance with exp.gdf.geometry containing (multi-)polygons
+        and/or (multi-)lines
+    impf_set : ImpactFuncSet
+        The set of impact functions.
+    haz : Hazard
+        The hazard instance.
+    res : float
+        Resolution of the disaggregation grid.
+    to_meters : bool, optional
+       If True, the geometries are projected to an equal area projection before
+       the disaggregation. res is then in meters. The exposures are
+       then reprojected into the original projections before the impact
+       calculation. The default is False.
+    disagg : string, optional
+        if 'avg' average value over points
+        if 'surf' value per points is equal to area per point (res*res)
+            for polygons, and distance per point (res) for lines.
+        if 'None' value is unchanged (or 1 if no value is defined)
+        Disaggregation method. The default is None.
+    agg : string, optional
+        If 'agg', the impact is averaged over all points in each geometry.
+        If 'sum', the impact is summed over all points in each geometry.
+        The default is 'sum'.
+
+    Returns
+    -------
+    Impact
+        Impact object with the impact per geometry (lines of exp.gdf).
+
+    See Also
+    --------
+    exp_geom_to_pnt: disaggregate exposures
+
+    """
+
+    #discretize exposure
+    exp_pnt = exp_geom_to_pnt(
+        exp=exp, res=res,
+        to_meters=to_meters, disagg=disagg
+        )
+    exp_pnt.assign_centroids(haz)
+
+    return calc_impact_pnt_agg(exp_pnt=exp_pnt, impf_set=impf_set, haz=haz, agg=agg)
```

Comment:
I'd consider refactoring a bit differently. ATM, there are three functions that have as output the aggregated impact per geometry [calc_geom_impact --> calc_impact_pnt_agg --> impact_pnt_agg]. Especially the first two funcs basically only differ in signature in that the first contains the original exposure and the second the disaggregated exposure, so I'd suggest merging the second into the first

Reply:
Could be an idea. I decided to do more methods to account for the different signatures. Thus, a user can choose to implement its own disaggregation method, and still use the calc impact. That was the idea to have first and second. What do you think?

Reply:
Right, that's a valid point. On the other hand, one question would be how realistic it is that someone does a non-regular disaggregation of the exposure that is not offered within  the `exp_geom_to_pnt` function, while still providing a `exp_pnt` in the next step that fulfills the structural requirements of that variable (e.g. a multiindex, a geometry column, an original geometry column, ..).
The advantage of having one wrapper function that does it all, is that it shows the three crucial steps immediately and comprehensively, i.e. exposure-disaggregation, classic impact calc, impact re-aggregation.

Reply:
But what is the downside of having an extra method? There is a method that does all in one already...

Reply:
Readability. At the moment, the overall function `calc_geom_impact` returns `calc_impact_pnt_agg` returns `impact_pnt_agg`. So one needs to scroll through more functions to get the main line of what the overall wrapper does.

Reply:
Readability for developers? This might be true. But I would say the use of the methods is there. So, maybe choosing clearer names would be a solution?